### PR TITLE
Kill Travis and Enable Bors

### PR DESCRIPTION
**Pull request by cevich**
***Depends on:***  #1848 #1971 #2036 #2121

### CODE REVIEW

**Comment by cevich** [comment]
@nalind @vrothberg [We've made great progress here on the ubuntu side.  Latest results show](https://cirrus-ci.com/build/5849523771670528) lots of 'expected cli option doesn't exist' kind of errors, as well as...did I mention I don't like using gpg in automation :grin:   What's my next step/fix needed here?

**Comment by cevich** [comment]
(rebased to include #2120)

**Comment by cevich** [comment]
(rebased again)

**Comment by cevich** [comment]
Previous run: https://cirrus-ci.com/task/6074545161371648

**Comment by cevich** [comment]
OOoooooooooo....progress!  Might even be able to wrap up this week!

**Comment by cevich** [comment]
@vrothberg @nalind @edsantiago  Looks like the last item remaining for both Ubuntu and Fedora is the info test (same/similar error in both):

https://cirrus-ci.com/task/6532738312306688?command=integration_test#L1004-L1029

**Comment by edsantiago** [comment]
Submitted #2126 for this, hope it helps

**Comment by vrothberg** [comment]
> Submitted #2126 for this, hope it helps

Great work everybody! #2126 has been merged, so we might be green after another rebase :crossed_fingers:

**Comment by cevich** [comment]
Rebased, and fingers-crossed.  Assuming it's green, I do want to do some testing of the Bors support on this PR:

[ ] Verify `bors try` functions (simulated merge)
[ ] Verify Cirrus-CI works behaves predictably with `bors try`
[ ] Confirm that skipping the 'cross' and 'gate' tasks post-merge is "okay" with everyone; note, they're still required to pass during PR testing.
[ ] Begin work on PR to add some basic docs for bors

**Comment by cevich** [comment]
:clap: [all green, thanks everyone!](https://cirrus-ci.com/build/5911210843439104) :clap:

**Comment by cevich** [comment]
bors try

**Comment by vrothberg** [comment]
Awesome! @cevich, once you feel comfortable to merge this, please remove the `WIP:`.

**Comment by cevich** [comment]
@vrothberg yep will do.

**Comment by cevich** [comment]
bors try

**Comment by cevich** [comment]
okay, we're good to go here.  @rhatdan [the test-merge I did above log is here if you want to peek.](https://github.com/containers/buildah/runs/419687894).  If everything looks good, give it a `bors r+` and that should cause this (and #1848) to be tested and then merge (if green).  I'll start in on a docs-update PR for bors-ng.

**Comment by rhatdan** [comment]
@cevich is this ready to review and merge?

**Comment by cevich** [comment]
@rhatdan yes ready to go.

Note:  I'm not sure what the 'TIMED OUT' message is from bors, since clearly (above that) my 'bors try' passed all tests (green check-mark).  It's possible if you `bors r+` this, we will hit something similar.  But I consider this growing-pains, and will address as needed.

**Comment by cevich** [comment]
We can see status here: https://app.bors.tech/batches/45255 still says `running` but (from the green check-mark) it should be finished and merged.  So I suspect we're about to get a timeout error.

**Comment by edsantiago** [comment]
I get EPERM on the bors.tech link above, even after linking my github account.

**Comment by cevich** [comment]
@edsantiago might be because you're not set as an org. admin in github (where bors gets it's permissions).

Assuming we get another timeout error.  At least Bors behavior is consistent between 'try' and 'merge' :smile: 

My guess is that it doesn't like the non 'success' or 'failure' status coming from Cirrus-CI (I marked some tasks as failure-tolerant).  After lunch I'll try removing that setting in `.cirrus.yml` but leave those names as optional for the bors status checks (in `bors.toml`).  That will bring back red-X marks next to some jobs, but maybe bors will (correctly) ignore them as configured.

**Comment by cevich** [comment]
bors try

**Comment by cevich** [comment]
:laughing: always one step forward and two steps back :laughing: 

This seems like a flake of some sort.  Going to rebase and try again.

**Comment by cevich** [comment]
note to me: investigate why bors try reported success but only listed 'cirrus-ci/required/in_podman'.

**Comment by cevich** [comment]
bors try

**Comment by cevich** [comment]
bors try-

**Comment by cevich** [comment]
bors try

**Comment by cevich** [comment]
Seems the fedora repository is flaking on us

**Comment by cevich** [comment]
bors try

**Comment by cevich** [comment]
Nnooooooooooooo :cry:

**Comment by cevich** [comment]
@rhatdan I was just reviewing the 'bors try' result (notice the red X).  There's a problem, and I wonder if we can even use bors-ng :disappointed:  It's not considering the status of ***ALL*** the `cirrus-ci/required/*` tests, it's picking one that passed and saying "good enough".  I doubt that's what we want :confounded:

**Comment by cevich** [comment]
Sorry, this is the 'bors try' I'm concerned about: https://github.com/containers/buildah/runs/425890269

Notice after that it posted a comment:  Build succeeded

**Comment by rhatdan** [comment]
Ok, sorry about that, I thought we were finally done.  I want travis gone ASAP since it is holding all of the PRs up waiting for it to decide to start.

**Comment by cevich** [comment]
No it's my fault as well, I should have immediately updated to WIP.

Yeah, you and me both, I was really hoping we were good to go here :cry:

**Comment by cevich** [comment]
Damnit! `*sigh*` okay.  I'm willing to take this to the upstream bors-ng community, maybe it's a bug, maybe the `bors.toml` is miss-configured.  it's not the end of the world.  It's better to find this problem early rather than late.

File: bors.toml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+# /#list-statuses-for-a-specific-ref) AND newer 'checks API'. Ref: https://developer.github.com/v3/checks
+# /runs/#list-check-runs-in-a-check-suite both return JSON:
+#
+# Status API:  Matches against '[].context' values
+# Checks API:  Matches against 'check_runs[].name' values
+#
+# Note: The wild-card character '%' is available.
+status = [
+    "cirrus-ci/required/in_podman%",
+    "cirrus-ci/required/testing%",
```
Comment (cevich) [inline bors.toml]: This is the line in question.  I was expecting that should match ***ALL*** of the testing tasks, but it only considered the first one.

**Comment by cevich** [comment]
...oh, I've got an idea.  Since it only wants to pick from a limited set of required status, we could do the same as we do on libpod: Have a 'success' task which depends on everything that's required.  That should be good-enough for now, and let this PR get merged while I followup with upstream bors-ng.

File: bors.toml (hunk 1 above)
Comment (edsantiago) [reply bors.toml]: Confirmed. The [docs I found](https://bors.tech/documentation/) explicitly say so:

>Patterns are written using SQL LIKE syntax. For example, this status directive will match **any** CI job that starts with ci/gitlab/:
>
>    status = [ 'ci/gitlab/%' ]

(emphasis mine)

**Comment by cevich** [comment]
Sorry @rhatdan for all the "almost" and "one more things" with this.  Automation can be very unforgiving, and so I'd rather err on the side of diligence when it comes to code-merging by a robot-overlord :grin: 

I think the ***last*** thing any maintainer wants is for a "known-failing" change to get merged in because the robot wasn't paying attention correctly.  Even if it was just a flake (above), I'm glad nobody needed to discover this bug  3,000 commits down the road :disappointed_relieved:

I think we've got it sorted now with a proven workaround.  I'll give it one-more 'bors try' to confirm it's watching the 'success' status now...

**Comment by cevich** [comment]
bors try

**Comment by cevich** [comment]
okay, good.  I think the root-cause was simply a flake, but bors is doing the right thing now.  I'm going to un-block and attempt to have bors merge it, under my watchful pair of vision-orbs.

**Comment by cevich** [comment]
ooohhh, someone gave you some smarts!

**Comment by cevich** [comment]
bors try

**Comment by cevich** [comment]
@rhatdan okay, I think I have the STOP-SHIP problem fixed now.  Though as a heads-up, I'm seeing some flakes during container-image pulls.  As per usual, these cause much :confounded: and there's about nothing we can do other than re-run :disappointed:   I'm presently babysitting re-runs on the failures above, otherwise (at last) this is actually ready to merge.  It's up to you if you want to wait for all-green + no flakes or not.

**Comment by cevich** [comment]
Hrmmm :sweat: damn...I think this might be an actual failure (it's not clearing by a re-run): 

https://cirrus-ci.com/task/5183589830623232?command=integration_test#L961-L976

@edsantiago have you seen ^^^^ before?  It didn't seem to be a problem yesterday :confused:

**Comment by edsantiago** [comment]
Looks like a transient. Give it one more try. (I can pull that image locally).

**Comment by cevich** [comment]
Thanks @edsantiago and yes I see one of my re-runs did eventually pass. okay, will just keep smashing my forehead against the button.

**Comment by cevich** [comment]
bors try

**Comment by cevich** [comment]
This is ready to merge, though like I'm doing with the try, the tests need babysitting due to networking flakes.

File: bors.toml (hunk 1 above)
Comment (cevich) [reply bors.toml]: so def. a bug of some sort then.

**Comment by cevich** [comment]
bors try-

**Comment by cevich** [comment]
bors try

**Comment by rhatdan** [comment]
Love seeing all tests pass and Bors says failed...

**Comment by cevich** [comment]
> Love seeing all tests pass and Bors says failed...

IKR...flakes :disappointed:

**Comment by cevich** [comment]
bors try

**Comment by cevich** [comment]
bors try

**Comment by edsantiago** [comment]
Is there any visibility into bors? Any way to get it to include links, or verbose logs, or something to indicate why it thinks the build failed?

**Comment by cevich** [comment]
> something to indicate why it thinks the build failed?

Bors operates very simply, by merging queued PRs into either the 'trying' or 'staging' branch, then letting CI execute a post-merge (branch) run.  So the run can be interacted with simply by visiting the branch, going under 'commits', and clicking the status icon (red "X" or green check).   As a short-cut, you can click the status icon the right-side, next to the commit sha, on the line below where it says:

"bors bot added a commit that referenced this pull request ... ago"

**Comment by edsantiago** [comment]
Oh - that little message never even registered (for me) as anything but noise. Thank you!

**Comment by cevich** [comment]
@edsantiago Since BATS doesn't have a per-test flake-retry mechanism like ginkgo, I was thinking of adding one.  It's not a perfect fix, but should help, especially as more tests are added and/or the matrix expands.  Sound okay?

**Comment by edsantiago** [comment]
That's a super-complicated issue right now. Can you hold off for a little while? I've had a PR blocked now for months, #2036, that intends to solve this very issue. (It is blocked because right now the CI images don't install podman, they use, ugh, docker. I'm assuming that this will be fixed very soon, since we need podman to test on f31. As soon as that gets fixed, I can get back to my PR).

**Comment by cevich** [comment]
> they use, ugh, docker

Oh!  Wait!  No!  ***THIS*** PR fixes that, it's all podman here now.  Maybe I can rebase this onto #2036?

**Comment by edsantiago** [comment]
Yikes - I don't know, doesn't that get circular? TBH I don't even remember the state of my PR right now - it was passing in my VMs, but that was a long time ago. It definitely needs to be rebased and updated. Let me look into that right now.

**Comment by cevich** [comment]
It won't get circular if it's rebased, then I rebase this onto it.  When this PR merges, it will drag the others along with it all at once.

**Comment by cevich** [comment]
Within this PR, the commit-list will look awful, but it works out after merging.

**Comment by cevich** [comment]
Let's see if moving that whalesay image over to quay helps.

**Comment by edsantiago** [comment]
To address the broader question: @nalind there is a hodgepodge of images in `digest.bats`, from various different registries. Is it a deliberate choice to hit on many different registries? If not, could some of those be cleaned up so we are more robust in the face of registry failures? TIA.

**Comment by cevich** [comment]
@edsantiago good question.  @nalind my last commit adds quay to `registries.conf`, and I've manually copied whalesay over there.  Same for 'fedora-minimal', since docker.io also seems to be randomly breaking :confounded:

**Comment by nalind** [comment]
> To address the broader question: @nalind there is a hodgepodge of images in `digest.bats`, from various different registries. Is it a deliberate choice to hit on many different registries? If not, could some of those be cleaned up so we are more robust in the face of registry failures? TIA.

Hitting multiple registries isn't a goal there - we're using images in registries that we know have manifests in specific formats, to check that images in each of those formats can be successfully pulled and used as base images, and that the commit logic doesn't get tripped up by differences between them.

**Comment by cevich** [comment]
> we know have manifests in specific formats

@nalind does that imply the tests depend on hard-coded references to metadata inside external content that we don't control?

Networking reliability aside, that makes me uncomfortable as it seems test-breakage would be both eventually-likely and unpredictable.

**Comment by edsantiago** [comment]
Back to the question I was trying to ask: @nalind is there any way we can change the list of images in this test, to (1) get more reliable registries (or at least not have to hit multiple ones), and (2) use smaller images, because the whale one is gigantic and hence more prone to network/disk failures?

**Comment by edsantiago** [comment]
@cevich I've pushed a new https://github.com/containers/buildah/pull/2036 and have some confidence in it, but I don't think it will help with the whalesay problem.

**Comment by cevich** [comment]
Thanks @edsantiago taking a look to see how to integrate it here.

**Comment by cevich** [comment]
okay, shuffled some things around, now #2036 should be running...lessee whah happn...

**Comment by nalind** [comment]
> @nalind does that imply the tests depend on hard-coded references to metadata inside external content that we don't control?
> 
> Networking reliability aside, that makes me uncomfortable as it seems test-breakage would be both eventually-likely and unpredictable.

For references that use tags, yes, different images could be pushed to those tags, and that would almost certainly change their manifests.  I'm hoping that doesn't happen with these tags, which refer to older versions of the images.  References that use digests can't be changed out from under us, because the result would be an image whose reference would include a different digest.

> Back to the question I was trying to ask: @nalind is there any way we can change the list of images in this test, to (1) get more reliable
 registries (or at least not have to hit multiple ones), and (2) use smaller images, because the whale one is gigantic and hence more prone to network/disk failures?

To do that, we'd need to change the references to point to other images whose manifests exhibit the characteristics we're looking for.  Aside from that requirement, I'm not picky about which ones we use.  These were merely cases that I was aware of (IIRC, because they were mentioned in bug reports).

**Comment by edsantiago** [comment]
@nalind what are those characteristics? The tests in `digest.bats` don't actually seem to check for anything at all AFAICT.

**Comment by cevich** [comment]
>point to other images whose manifests exhibit the characteristics we're looking for. 

Thanks @nalind for explaining / confirming.

> hoping that doesn't happen

There's no forgiveness of in automation, even with infinite hope available :cry:  We've really got to exert control over both ends.  I'm happy to shove these special images somewhere static, where they can "never" be touched.  However, even that seems slightly fragile.  Is there a way to define these images + specific manifests in code, so they can be re-created if need-be?

**Comment by edsantiago** [comment]
Can we just take a step back and figure out what exactly is being tested? AFAICT the digest test simply does a pull-from-remote followed by a number of local push/commit steps. It doesn't actually check anything. @nalind if you help me understand what the purpose of this test is, I could try to (a) come up with better, safer, more representative images, and then (b) rewrite the test so it actually tests something.

**Comment by cevich** [comment]
@edsantiago we seem to be in much better shape now including your #2036 in the mix.  The only thing I'm seeing is that looks new is: https://cirrus-ci.com/task/5267044266409984?command=integration_test#L360

My intention was to clobber F29 here, once we got F31 up-n-going in libpod.  However, I'm not opposed to disabling it preemptively if that failure is a PITA to fix (the test seems to be working fine elsewhere).

**Comment by cevich** [comment]
> purpose of this test is, I could try to

Yeah!  That's what I was trying to say :smile:

**Comment by edsantiago** [comment]
> The only thing I'm seeing is that looks new is [...]
>[+1481s] not ok 286 images json test
>[+1481s] # #| Expected 30 lines of output, got 26

Interesting. It looks like `history` is getting lost. That's not happening on my test setup. Looking into it further now...

**Comment by edsantiago** [comment]
Root cause: too-old podman. 1.6.2 does not seem to handle image history.

I despise number-of-lines tests; so I'll try to rewrite it to something better. Sadly, we don't have `jq` in our test images, so I won't write anything that relies on it. But as a low-priority pretty-please, for some day, could you maybe add `jq` into the CI images so future tests can use it? TIA

**Comment by nalind** [comment]
> @nalind what are those characteristics? The tests in `digest.bats` don't actually seem to check for anything at all AFAICT.

We're checking that a given image can be pulled, used as a base image for a build, and that a new image that we build using that pulled image as a base can successfully be pushed to somewhere other than local storage, when the image is:
* a v2s1 image referenced using its digest
* a v2s1 image where one of the entries is a duplicate that gets discarded when it's converted to v2s2
* a v2s1 image referenced using its tag
* a v2s2 image referenced using its digest (incorrectly - this is also a v2s1 image, `alpine@sha256:25fd8fc1aefcc8ae46aae23daefcd7dcb97f676fa0bc72bb0bf7cfb75df4f22e` is a better candidate)
* a v2s2 image referenced using its tag (incorrectly - this is the same v2s1 image, `alpine:3.1` is a better candidate, though it's a manifest list now, so the test won't pass on non-amd64 systems without tweaks.
* a v2s2 image referenced using its repository name, with neither a tag nor a digest.  It's a manifest list, so the test won't pass on non-amd64 systems without tweaks.

**Comment by cevich** [comment]
>   could you maybe add jq into the CI images

Oh, that's actually really easy to do in buildah-CI since we're not caching VM images (they're built from scratch every damn time)...

**Comment by cevich** [comment]
@edsantiago done...added `jq` and disabled F29.

**Comment by edsantiago** [comment]
@cevich , new #2036 force-pushed, should handle the `images json` case.

@nalind, thank you, I'm trying to (ahem) digest all that. IIUC you are saying that those images are all hand-picked due to [certain characteristics], and that the test doesn't need to check those characteristics, it simply needs to guarantee that buildah doesn't crash when operating on them? I have one big question though:
> can successfully be pushed to somewhere other than local storage

AFAICT `digest.bats:fromreftest()` only pushes to local storage. Is that an oversight, or have I missed something?

**Comment by edsantiago** [comment]
@cevich, thanks, our messages crossed in the ether. If you disabled f29, you should be able to rerun tests and have them pass (with or without rebaseing on my new force-push). Thank you!

**Comment by cevich** [comment]
Yep, rebasing anyway just to be sure since this is now a franken-PR :smile:

**Comment by nalind** [comment]
> @nalind, thank you, I'm trying to (ahem) digest all that. IIUC you are saying that those images are all hand-picked due to [certain characteristics], and that the test doesn't need to check those characteristics, it simply needs to guarantee that buildah doesn't crash when operating on them?

That, plus we make sure that we generate an image that can be pushed.  Some of the bug reports noted that we succeeded at committing a new image, but subsequently failed to push that image when logic in containers/image noticed, or was tripped up by, an error in how we'd constructed the data that we had generated at commit-time.

> I have one big question though:
> 
> > can successfully be pushed to somewhere other than local storage
> 
> AFAICT `digest.bats:fromreftest()` only pushes to local storage. Is that an oversight, or have I missed something?

Sorry, I was being imprecise.  By "local storage", I mean the storage managed by containers/storage.  For the purposes of testing the committing and pushing bits that these images previously tripped up (i.e., the logic that builds the new image's config blob and manifest), a `dir:` destination works as well as a `docker://` destination would, since in both cases we end up using the containers/image image-copying logic.

**Comment by edsantiago** [comment]
@nalind thank you, that helps me understand. Since tests are green (!) (!!!!), I'd like to move discussion of this aspect over to #2147.

**Comment by rhatdan** [comment]
@edsantiago @cevich @nalind Does that mean this PR is ready to merge?

**Comment by edsantiago** [comment]
@rhatdan I sure hope so but let's wait for @cevich to supervise the process

**Comment by cevich** [comment]
yes please.  hurry before the robots realize it's Friday!

**Comment by cevich** [comment]
Big thanks @edsantiago @nalind @rhatdan for help and patients with this effort.  Hopefully time will prove it was a worthwhilen investment..

**Comment by rhatdan** [comment]
:+1:

**Comment by cevich** [comment]
Well...the good news is I did ***NOT*** need to do a single re-run of any test.  I believe that message from Bors (above) indicates that master changed during execution.  That could be caused by a non-bors entity merging a PR or otherwise changing the target branch HEAD.

**Comment by cevich** [comment]
bors retry

**Comment by cevich** [comment]
(timeout building a container image in a test)

**Comment by cevich** [comment]
I opened https://forum.bors.tech/t/bug-wildcard-status-ignores-1-match/438 regarding the wildcard false-positive problem discovered back here https://github.com/containers/buildah/pull/2115#issuecomment-582069179

**Pull request merged**